docs/guidelines.md

```diff
@@ -0,0 +1,200 @@
+
+For context-specific jobs (e.g. when using matrix), consider using parentheses:
+
+- `Build (${{ env.GO_VERSION }})` -> `Go / Build (1.24.2)`
+- `Deploy (staging)`
+
+### 4.5. Steps
+
+Jobs consist of steps, each of which should be:
+
+- Clear in intent.
+- Named descriptively.
+- Ordered logically.
+- Commented if they contain non-obvious logic.
+
+It is good practice to use step names that clearly describe their function, improving maintainability and readability.
+
+Step keys should be ordered as:
+
+- `name`
+- `id`
+- `if`
+- `shell`
+- `working-directory`
+- `env`
+- `uses` / `run`
+- `with`
+
+Ordering keys consistently improves readability and reduces unnecessary diffs during reviews.
+
+```yaml
+- name: "Checkout repository"
+  uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683
+  with:
+    path: "./example/"
+
+- name: "Read version"
+  id: version
+  run: |
+    VERSION=$(cat version)
+    echo "version=$VERSION" >> "$GITHUB_OUTPUT"
+
+- name: "Install dependencies"
+  working-directory: "example/"
+  env:
+    GO_VERSION: "1.24.x"
+    TEST_VERSION: "${{ steps.version.outputs.version }}"
+  run: make deps
+```
+
+## 5. Actions
+
+- Repository-specific actions: `.github/actions/`
+- Reusable actions: `hemilabs/actions/`
+
+Refer to [Reusable actions and workflows](#1-reusable-actions-and-workflows) for reusable actions.
+
+### 5.1. Composite Actions
+
+When multiple steps are commonly reused, it may be helpful to bundle them into a composite action.
+
+- Prefer composite actions for reusing workflow logic.
+- Keep composite actions minimal and focused on a single responsibility, for example, installing a tool or setting up an
+  environment.
+- Document inputs, outputs, and usage clearly in a `README.md` file.
+
+### 5.2. Custom Actions
+
+When it is necessary to implement functionality beyond what is possible in a workflow or composite action, a custom
+JavaScript or Docker-based action could be created.
+
+Custom actions introduce more maintenance overhead and complexity (especially Docker-based). Use only when composite
+actions are insufficient.
+
+## 6. Collaboration
+
+To encourage clarity and consistency across all actions and workflows:
+
+- **Propose early, discuss openly** - When introducing or updating shared actions or workflows, open a pull request
+  early and use it as a space for open discussion.
+- **Align on patterns** - Where multiple valid approaches exist, follow established patterns unless there's a strong
+  and documented reason to diverge.
+- **Add documentation** - Ensure changes and usage are well-documented, making it easy for teams to understand what
+  changed, why it changed, and how to adopt it.
+
+These guidelines are intended to support consistency and maintainability across teams. Feedback is welcome, and
+improvements to these guidelines are always encouraged.
+
+To suggest changes to these guidelines, feel free to open an issue, pull request, or send a message in Slack.
```

Comment:
Since these are public docs, I would remove the reference to Slack as it is internal only.